### ISSUE TITLE
[no-unused-vars] does not work if type is only used in union in generic parameter

### ISSUE THREAD

**Issue by SevInf** [bradzacher/eslint-plugin-typescript#218]
**Repro**

```JSON
{
        "no-unused-vars": "error",
        "typescript/no-unused-vars": "error"
    }
}
```

```TS
import { SomeType, SomeOtherType } from './types';

export function returnsPromise(): Promise<SomeType | SomeOtherType> {
}

```

**Expected Result**

`SomeType` and `SomeOtherType` correctly marked as used.

**Actual Result**

```
  1:10  error  'SomeType' is defined but never used       no-unused-vars
  1:20  error  'SomeOtherType' is defined but never used  no-unused-vars
```

**Additional Info**

Works as expected if I return `SomeType | SomeOtherType` instead of `Promise` or declare the types in the same file or return single type instead of union.

**Versions**

| package                    | version |
| -------------------------- | ------- |
| `eslint-plugin-typescript` | `0.14.0` |
| `typescript-eslint-parser` | `21.0.2` |
| `typescript`               | `3.1.6` |

**Comment by bradzacher** [comment]
Please try using `typescript-eslint-parser@^17`
This will be fixed as of 1.0.0 release
